### PR TITLE
Write ELLPACK pages to disk

### PR DESCRIPTION
**Pull request by rongou**
Whew, that took a while. The compressed ELLPACK buffers are written to disk now. I haven't changed the `gpu_hist` updater to handle multiple pages yet. Will work on that next.

Part of #4357.

@RAMitchell @trivialfis @sriramch

### CODE REVIEW

**Comment by trams** [comment]
Thank you for your contribution.
Could you provide some context on what kind of issue are you trying to solve? I am a bit lost (granted I am not very familiar with GPU specifics and pull requests seems to address this side of xgboost)

**Comment by rongou** [comment]
As the title indicates, this is a work in progress, not quite ready for review yet. :) Probably still a couple commits away from being fully functional. In any case, it's part of the effort to support external memory on GPUs. See #4357.

**Comment by hcho3** [comment]
@rongou It looks like one of the Docker servers is temporarily rejecting the traffic from our CI server. Would you wait a little bit and try later?

Alternatively, are there things that you'd like to test that can be only run in the CI server? I'd like to improve the doc so that developers can run most tests locally.

**Comment by hcho3** [comment]
@rongou Aside: I'm upgrading GPU workers to G4 this week.

**Comment by rongou** [comment]
@hcho3 The main thing is I don't have access to a windows dev machine. The windows compiler seems quite old and doesn't like some of the c++11 features. Anyway, I think I'm getting very close. I'll close this PR and open a new one when it's ready.

**Comment by hcho3** [comment]
@rongou Ah got it. Yes, Windows is kind of special and harder to manage as part of DevOps. It would be really nice if NVIDIA docker supported Windows, but for now I manually manage Windows AMIs and install software packages by hand.

**Comment by hcho3** [comment]
@rongou As for C++11 support in Visual Studio, we are using Visual Studio 2017, so C++11 should be fully supported (according to [this link](https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019))

**Comment by hcho3** [comment]
@rongou Thanks for your patience. I fixed issues in the CI in #4921 and now re-started the tests.

**Comment by rongou** [comment]
@trivialfis I added code comments. A few implementation notes:
* In external memory mode, we first loop through all the CSR pages to build the quantile sketch and create an `EllpackInfo` to hold that information. Then we loop through the CSR pages again, this time compressing each CSR page into ELLPACK. The resulting compressed buffer is copied back to host and appended together. When the size reaches a threshold, the page is written to disk.
* The `EllpackInfo` is actually kept in memory. When reading back the ellpack pages from disk, only the compressed buffers are read in and copied to GPU.
* A few key methods are probably `EllpackPageImpl::Push`, `EllpackPageImpl::InitDevice`, `EllpackPageSourceImpl` constructor, and `EllpackPageSourceImpl::WriteEllpackPages`.

Hope this helps. Thanks!

**Comment by trivialfis** [comment]
Will review soon.  Thanks for the brief introduction.

File: src/data/ellpack_page.cu [hunk 1]
```diff
@@ -185,13 +195,82 @@ void EllpackPageImpl::CreateHistIndices(int device,
...
+// Copy the compressed features to GPU.
+void EllpackPageImpl::InitDevice(int device, EllpackInfo info) {
+  if (device_initialized_) return;
+
+  monitor_.StartCuda("CopyPageToDevice");
+  dh::safe_cuda(cudaSetDevice(device));
+
+  gidx_buffer = {};
+  ba_.Allocate(device, &gidx_buffer, idx_buffer.size());
+  dh::CopyVectorToDeviceSpan(gidx_buffer, idx_buffer);
```
Comment (sriramch) [inline src/data/ellpack_page.cu]: so, in this version the sparse page iterator copies the entire compressed buffer into gpu and returns?

File: src/data/ellpack_page_source.cu [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+const EllpackPage& EllpackPageSourceImpl::Value() const {
+  EllpackPage& page = source_->Value();
+  page.Impl()->InitDevice(device_, ellpack_info_);
+  return page;
+}
+
+// Compress each CSR page to ELLPACK, and write the accumulated pages to disk.
+void EllpackPageSourceImpl::WriteEllpackPages(DMatrix* dmat, const std::string& cache_info) const {
+  auto cinfo = ParseCacheInfo(cache_info, kPageType_);
+  SparsePageWriter<EllpackPage> writer(cinfo.name_shards, cinfo.format_shards, 6);
```
Comment (sriramch) [inline src/data/ellpack_page_source.cu]: can literal 6 be defined before to denote what it means?

File: src/data/ellpack_page_source.cu [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  double tstart = dmlc::GetTime();
+  for (const auto& batch : dmat->GetBatches<SparsePage>()) {
+    impl->Push(device_, batch);
+
+    if (impl->MemCostBytes() >= DMatrix::kPageSize) {
+      bytes_write += impl->MemCostBytes();
+      writer.PushWrite(std::move(page));
+      writer.Alloc(&page);
+      impl = page->Impl();
+      impl->matrix.info = ellpack_info_;
```
Comment (sriramch) [inline src/data/ellpack_page_source.cu]: is the ellpack info also written with the compressed data for every page? if a dataset contains large number of features, wouldn't this cost be prohibitive across a large number of such pages?

File: src/data/sparse_page_source.h [hunk 4]
```diff
@@ -230,30 +264,19 @@ class SparsePageSource : public DataSource<T> {
...
-    LOG(INFO) << "SparsePageSource::CreateRowPage Finished writing to "
-              << name_info;
+    LOG(INFO) << "SparsePageSource::CreateRowPage Finished writing to " << cinfo.name_info;
   }
 
   /*!
    * \brief Create source cache by copy content from DMatrix.
-   * \param cache_info The cache_info of cache file location.
-   */
-  static void CreateRowPage(DMatrix* src,
```
Comment (sriramch) [inline src/data/sparse_page_source.h]: why is the row page creation removed? is it renamed as something - cached row page?

File: include/xgboost/data.h [hunk 5]
```diff
@@ -242,13 +262,6 @@ class SparsePage {
    * \param batch The row batch to be pushed
    */
   void PushCSC(const SparsePage& batch);
-  /*!
-   * \brief Push one instance into page
-   *  \param inst an instance row
-   */
-  void Push(const Inst &inst);
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/data.h]: +1

File: src/data/ellpack_page.cu [hunk 6]
```diff
@@ -12,14 +10,23 @@
...
+  return impl_->Size();
+}
+
+void EllpackPage::SetBaseRowId(size_t row_id) {
+  impl_->SetBaseRowId(row_id);
+}
 
 // Bin each input data entry, store the bin indices in compressed form.
-template<typename std::enable_if<true,  int>::type = 0>
+template<typename std::enable_if<true, int>::type = 0>
```
Comment (trivialfis) [inline src/data/ellpack_page.cu]: Actually why is this line needed?

File: src/data/ellpack_page.cuh [hunk 7]
```diff
@@ -42,56 +40,68 @@ __forceinline__ __device__ int BinarySearchRow(
...
+  /*! \brief Row length for ELLPack, equal to number of features. */
+  size_t row_stride;
+  /*! \brief Total number of bins, also used as the null index value, . */
+  size_t n_bins;
+  /*! \brief Minimum value for each feature. Size equals to number of features. */
+  common::Span<bst_float> min_fvalue;
+  /*! \brief Histogram cut pointers. Size equals to (number of features + 1). */
+  common::Span<uint32_t> feature_segments;
+  /*! \brief Histogram cut values. Size equals to (bins per feature * number of features). */
+  common::Span<bst_float> gidx_fvalue_map;
```
Comment (trivialfis) [inline src/data/ellpack_page.cuh]: These are not trivial when dataset is wide.  But we can consider optimization after having a working pipeline.

File: src/data/ellpack_page_source.cu (hunk 3 above)
Comment (trivialfis) [reply src/data/ellpack_page_source.cu]: > is the ellpack info also written with the compressed data for every page?

@sriramch I don't think it's written to disk for every page according to `EllpackPageRawFormat::Write`.  @rongou Correct me if i'm wrong.

File: src/data/sparse_page_writer.h [hunk 8]
```diff
@@ -131,8 +190,21 @@ struct SparsePageFormatReg
  *   });
  * \endcode
  */
+#define SparsePageFmt SparsePageFormat<SparsePage>
```
Comment (trivialfis) [inline src/data/sparse_page_writer.h]: The definition is out of necessity or convenience?

File: tests/cpp/data/test_sparse_page_dmatrix.cu [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+namespace xgboost {
+
+TEST(GPUSparsePageDMatrix, EllpackPage) {
+  dmlc::TemporaryDirectory tempdir;
+  const std::string tmp_file = tempdir.path + "/simple.libsvm";
+  CreateSimpleTestData(tmp_file);
+  DMatrix* dmat = DMatrix::Load(tmp_file + "#" + tmp_file + ".cache", true, false);
+
+  // Loop over the batches and assert the data is as expected
+  for (const auto& batch : dmat->GetBatches<EllpackPage>({0, 256, 64})) {
```
Comment (trivialfis) [inline tests/cpp/data/test_sparse_page_dmatrix.cu]: Sorry for the nitpick, what if the parameter is changed, it might happen between runs.

File: src/data/sparse_page_writer.h [hunk 10]
```diff
@@ -23,9 +23,14 @@
...
 namespace xgboost {
 namespace data {
+
+template<typename T>
+struct SparsePageFormatReg;
+
 /*!
  * \brief Format specification of SparsePage.
  */
+template<typename T>
```
Comment (trivialfis) [inline src/data/sparse_page_writer.h]: Give a brief for this T in above doc please.

File: src/data/sparse_page_source.h [hunk 11]
```diff
@@ -312,9 +327,7 @@ class SparsePageSource : public DataSource<T> {
       size_t bytes_write = 0;
       double tstart = dmlc::GetTime();
       for (auto& batch : src->GetBatches<SparsePage>()) {
-        if (page_type == ".row.page") {
```
Comment (trivialfis) [inline src/data/sparse_page_source.h]: Do we still have `row.page`?

File: src/data/sparse_page_source.h [hunk 12]
```diff
@@ -338,28 +351,22 @@ class SparsePageSource : public DataSource<T> {
       if (page->data.Size() != 0) {
         writer.PushWrite(std::move(page));
       }
-
-      std::unique_ptr<dmlc::Stream> fo(
-          dmlc::Stream::Create(name_info.c_str(), "w"));
-      int tmagic = kMagic;
-      fo->Write(&tmagic, sizeof(tmagic));
-      info.SaveBinary(fo.get());
```
Comment (trivialfis) [inline src/data/sparse_page_source.h]: Where did the meta info go?

File: src/data/ellpack_page.cu (hunk 6 above)
Comment (rongou) [reply src/data/ellpack_page.cu]: Good question. :) I read this page: https://eli.thegreenplace.net/2014/sfinae-and-enable_if/, but still don't understand why it's here. Removed.

File: src/data/ellpack_page.cu [hunk 13]
```diff
@@ -185,13 +195,82 @@ void EllpackPageImpl::CreateHistIndices(int device,
...
+  monitor_.StopCuda("BinningCompression");
+
+  monitor_.StartCuda("CopyDeviceToHost");
+  std::vector<common::CompressedByteT> buffer(gidx_buffer.size());
+  dh::CopyDeviceSpanToVector(&buffer, gidx_buffer);
+  int offset = 0;
+  if (!idx_buffer.empty()) {
+    offset = ::xgboost::common::detail::kPadding;
+  }
+  idx_buffer.reserve(idx_buffer.size() + buffer.size() - offset);
```
[the remark this answers is not in the document]
Comment (rongou) [reply src/data/ellpack_page.cu]: We know the size of the dataset, but we don't want to store the whole dataset in a single page. Since we are fed CSR pages one by one, I don't think we know beforehand how big the buffer is going to be.

File: src/data/ellpack_page.cu (hunk 1 above)
Comment (rongou) [reply src/data/ellpack_page.cu]: Yes.

File: src/data/ellpack_page.cuh (hunk 7 above)
Comment (rongou) [reply src/data/ellpack_page.cuh]: Agreed. It's actually kept in memory.

File: src/data/ellpack_page_source.cu [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+    return fi->Read(&impl->idx_buffer);
+  }
+
+  bool Read(EllpackPage* page,
+            dmlc::SeekStream* fi,
+            const std::vector<bst_uint>& sorted_index_set) override {
+    auto* impl = page->Impl();
+    if (!fi->Read(&impl->n_rows))  return false;
+    return fi->Read(&page->Impl()->idx_buffer);
+  }
```
[the remark this answers is not in the document]
Comment (rongou) [reply src/data/ellpack_page_source.cu]: Done.

File: src/data/ellpack_page_source.cu (hunk 2 above)
Comment (rongou) [reply src/data/ellpack_page_source.cu]: Done.

File: src/data/ellpack_page_source.cu (hunk 3 above)
Comment (rongou) [reply src/data/ellpack_page_source.cu]: Right, it's kept in memory.

File: src/data/sparse_page_source.h [hunk 15]
```diff
@@ -46,6 +46,47 @@ GetCacheShards(const std::string& cache_info) {
 
 namespace xgboost {
 namespace data {
+
+/*!
+ * \brief decide the format from cache prefix.
+ * \return pair of row format, column format type of the cache prefix.
+ */
+inline std::pair<std::string, std::string> DecideFormat(const std::string& cache_prefix) {
```
[the remark this answers is not in the document]
Comment (rongou) [reply src/data/sparse_page_source.h]: Mainly to avoid dealing generics.

File: src/data/sparse_page_source.h (hunk 4 above)
Comment (rongou) [reply src/data/sparse_page_source.h]: This version is not being used.

File: src/data/sparse_page_source.h (hunk 11 above)
Comment (rongou) [reply src/data/sparse_page_source.h]: There is another method `CreateRowPage()` that creates CSR pages.

File: src/data/sparse_page_source.h (hunk 12 above)
Comment (rongou) [reply src/data/sparse_page_source.h]: In the current flow, we create the row pages first and write the meta info there. This method now only creates column pages and sorted column pages from the existing row pages, so the meta info should be already written.

File: src/data/sparse_page_writer.h (hunk 10 above)
Comment (rongou) [reply src/data/sparse_page_writer.h]: Done.

File: src/data/sparse_page_writer.h (hunk 8 above)
Comment (rongou) [reply src/data/sparse_page_writer.h]: The `DMLC_REGISTRY_REGISTER` macro doesn't like angle brackets.

File: tests/cpp/data/test_sparse_page_dmatrix.cu (hunk 9 above)
Comment (rongou) [reply tests/cpp/data/test_sparse_page_dmatrix.cu]: Good catch. I saved `BatchParam` and added some checks.

File: src/data/ellpack_page.cu (hunk 1 above)
Comment (sriramch) [reply src/data/ellpack_page.cu]: in the future version then, will the `idx_buffer` go away and that the ellpack info also be persisted on disk in a separate page?

File: src/data/ellpack_page.cu (hunk 6 above)
Comment (sriramch) [reply src/data/ellpack_page.cu]: i think that was added to prevent multiple symbol definition linker error. some of the tests includes the cuda file to test internal methods, and this kernel will be exposed as a global function with external linkage in multiple translation units. when they are put together to build a binary, linker will invariably complain.

making it `inline` or a function template will make the symbol visible only internally (internal linkage). it *could* have been made static as well to force an internal linkage; not sure why a function template was chosen.

this is moot presently, as with the ellpack refactorization, the internal kernel is moved to a separate file that isn't explicitly/implicitly included within the tests.

File: src/data/ellpack_page.cu (hunk 1 above)
Comment (rongou) [reply src/data/ellpack_page.cu]: I think we'd always need the `idx_buffer` for reading from files. We might need to persist the ellpack info if we want to use the same binning on a different dataset.

File: src/data/sparse_page_source.h (hunk 4 above)
Comment (trivialfis) [reply src/data/sparse_page_source.h]: @rongou Thanks for cleaning up the code.

**Comment by trivialfis** [comment]
@RAMitchell Do you need to review?

**Comment by trivialfis** [comment]
Merging.  @RAMitchell If you have some ideas on how to refactor please do share.

**Comment by trivialfis** [comment]
@rongou Looking forward to your next PR. ;-)
